drafts/bitcartel-nu0-network-handshake/draft.rst

```diff
@@ -95,36 +95,62 @@ Network Coalescence
 
-The eviction code can be updated as follows::
+The eviction candidates can be modified as so::
 
     static bool AttemptToEvictConnection(bool fPreferNewConnection) {
     ...
     // Protect connections with certain characteristics
     ...
-    // Check version of eviction candidates
-    if (current_block_height >= (activationheight - NETWORK_COALESCE_BLOCK_PERIOD)) {
-      // if there exist any legacy nodes, keep them in the eviction set
-      // and at the same time remove overwinter nodes from eviction set.
-      // if there do not exist any legacy nodes,
-      // continue with existing behaviour.
+    // Check version of eviction candidates...
+    // If we are connected to any pre-Overwinter nodes, keep them in the eviction set and remove any Overwinter nodes
+    // If we are only connected to Overwinter nodes, continue with existing behaviour.
+    if ((height < nActivationHeight) &&
+        (height >= (nActivationHeight - NETWORK_UPGRADE_PEER_PREFERENCE_BLOCK_PERIOD)))
+    {
+        // Find any nodes which don't support Overwinter protocol version
+        BOOST_FOREACH(const CNodeRef &node, vEvictionCandidates) {
+            if (node->nVersion < OVERWINTER_PROTO_VERSION) {
+                vTmpEvictionCandidates.push_back(node);
+            }
+        }
 
+        // Prioritize these nodes by replacing eviction set with them
+        if (vTmpEvictionCandidates.size() > 0) {
+            vEvictionCandidates = vTmpEvictionCandidates;
+        }
+    }
 
-Disconnecting Existing Connections
-----------------------------------
+The existing method of disconnecting a candidate remains:
+
+    vEvictionCandidates[0]->fDisconnect = true;
+
+The existing eviction process also classifies and divides eviction candidates into groups.  If a group only has one peer, it will not be evicted.  This means at least one pre-Overwinter node should remain connected upto the activation block height, barring any network issues or a high ban score.
 
-It is likely that at the activation block height, an Overwinter node will still be connected to some Legacy nodes.
 
-Currently, when connecting, a node must perform the networking handshake, and send the version message, before any other messages are processed.
+Disconnecting Existing Connections
+----------------------------------
 
-To disconnect existing connections, we can modify ProcessMessage so that the protocol version is always checked after Overwinter activates.
+At the activation block height, an Overwinter node may still remain connected to pre-Overwinter nodes.  Currently, when connecting, a node can only perform the networking handshake once, where it sends the version message before any other messages are processed.  To disconnect existing pre-Overwinter connections, ProcessMessage is modified so that once Overwinter activates, the protocol version of a peer is always checked when processing inbound messages.
 
 Example code::
```

Comment:
This seems to complicate `processmessage` with tasks not implied by its name (disconnecting nodes)
might it be better, e.g., to scan for pre-overwinter nodes when activationheight is reached, and disconnect all of them?
Seems a shame to make these checks everytime ProcessMessage is called, as this is relevant only for nodes who were running during activation height. When the node restarts after activation height it won't connect to pre-overwinter nodes in the first place.

Reply:
Modified to clear this up and in code an extra conditional and reordering so the version check is performed only if necessary.  Once pure Overwinter nodes are available, we won't need this check.

Reply:
The extra condition you refer to is `MIN_PEER_PROTO_VERSION < OVERWINTER_PROTO_VERSION` ?
So in an upgrade after overwinter activation, we'll update the value of `MIN_PEER_PROTO_VERSION` to `OVERWINTER_PEER_PROTO_VERSION`?
If so, might mention this in the ZIP.